libfwupd/fwupd-device.c

```diff
@@ -1612,6 +1612,92 @@ fwupd_pad_kv_ups (GString *str, const gchar *key, FwupdUpdateState value)
 	fwupd_pad_kv_str (str, key, fwupd_update_state_to_string (value));
 }
 
+/**
+ * fwupd_device_to_json:
+ * @builder: A #JsonBuilder
+ * @device: A #FwupdDevice
+ *
+ * Adds a fwupd device to a JSON builder
+ *
+ * Since: 1.2.6
+ **/
+void
+fwupd_device_to_json (JsonBuilder *builder, FwupdDevice *device)
+{
+	FwupdDevicePrivate *priv = GET_PRIVATE (device);
+
+	g_return_if_fail (FWUPD_IS_DEVICE (device));
+	g_return_if_fail (builder != NULL);
+
+	json_builder_set_member_name (builder, FWUPD_RESULT_KEY_NAME);
+	json_builder_add_string_value (builder, priv->name);
+	json_builder_set_member_name (builder, FWUPD_RESULT_KEY_DEVICE_ID);
+	json_builder_add_string_value (builder, priv->id);
+	json_builder_set_member_name (builder, FWUPD_RESULT_KEY_PARENT_DEVICE_ID);
+	json_builder_add_string_value (builder, priv->parent_id);
+	for (guint i = 0; i < priv->guids->len; i++) {
+		const gchar *guid = g_ptr_array_index (priv->guids, i);
+		g_autofree gchar *title = g_strdup_printf ("%s%u", FWUPD_RESULT_KEY_GUID, i);
+		json_builder_set_member_name (builder, title);
+		json_builder_add_string_value (builder, guid);
+	}
+	json_builder_set_member_name (builder, FWUPD_RESULT_KEY_SERIAL);
+	json_builder_add_string_value (builder, priv->serial);
+	json_builder_set_member_name (builder, FWUPD_RESULT_KEY_SUMMARY);
+	json_builder_add_string_value (builder, priv->summary);
+	json_builder_set_member_name (builder, FWUPD_RESULT_KEY_DESCRIPTION);
+	json_builder_add_string_value (builder, priv->description);
+	json_builder_set_member_name (builder, FWUPD_RESULT_KEY_PLUGIN);
+	json_builder_add_string_value (builder, priv->plugin);
+	json_builder_set_member_name (builder, FWUPD_RESULT_KEY_FLAGS);
+	json_builder_add_int_value (builder, priv->flags);
+	for (guint i = 0; i < priv->checksums->len; i++) {
+		const gchar *checksum = g_ptr_array_index (priv->checksums, i);
+		g_autofree gchar *checksum_display = fwupd_checksum_format_for_display (checksum);
+		g_autofree gchar *title = g_strdup_printf ("%s%u", FWUPD_RESULT_KEY_CHECKSUM, i);
+		json_builder_set_member_name (builder, title);
+		json_builder_add_string_value (builder, checksum_display);
+	}
+	json_builder_set_member_name (builder, FWUPD_RESULT_KEY_VENDOR);
```

Comment:
should we include things that are NULL? e.g. maybe we should check `priv->vendor` before doing `json_builder_set_member_name()` and `json_builder_add_string_value()`

Reply:
I oscillated on this because it's a whole bunch of if `(tmp != NULL)` in the function.  I'll make a static helper function that does this.

Reply:
No need for tmp, the compiled will cache a single dereference.

Reply:
I was meaning it is less readable/ugly to call the same thing over and over.  helper will make it more readable.